### ISSUE TITLE
pyinstaller "FileNotFoundError", for files with special characters in them such as "[".

### ISSUE THREAD

**Issue by StuntsPT** [pyinstaller/pyinstaller#2206]
I am currently using a GTK3 theme called "Vertex", which contains some non-standard file-names, such as `/usr/share/themes/Firefox/Vertex/chrome/images/arrow-dn[dropmarker].png`
As you can see, the filename contains the "[" and "]" characters. This is causing pyinstaller to stop building the binary an throwing the following error:

``` python
Traceback (most recent call last):
  File "/usr/bin/pyinstaller2", line 9, in <module>
    load_entry_point('PyInstaller==3.2', 'console_scripts', 'pyinstaller2')()
  File "/usr/lib/python2.7/site-packages/PyInstaller/__main__.py", line 90, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/usr/lib/python2.7/site-packages/PyInstaller/__main__.py", line 46, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/build_main.py", line 788, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/build_main.py", line 734, in build
    exec(text, spec_namespace)
  File "<string>", line 16, in <module>
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/build_main.py", line 212, in __init__
    self.__postinit__()
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/datastruct.py", line 178, in __postinit__
    self.assemble()
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/build_main.py", line 470, in assemble
    module_hook.post_graph()
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/imphook.py", line 409, in post_graph
    self._load_hook_module()
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/imphook.py", line 390, in _load_hook_module
    attr_value = sanitizer_func(attr_value)
  File "/usr/lib/python2.7/site-packages/PyInstaller/building/utils.py", line 456, in format_binaries_and_datas
    src_root_path_or_glob))
PyInstaller.compat.FileNotFoundError: Path or glob "/usr/share/themes/Firefox/Vertex/chrome/images/arrow-dn[dropmarker].png" not found or matches no files.

```

The file is present in my filesystem.
I also have no idea why this file is even being pulled into the bundle, but that is another issue.

Edit: Added the entire traceback.

**Comment by tallforasmurf** [comment]
This would appear to be an issue with glob. At least under Mac OS:

    >>> glob.glob( "/Users/dcortes1/Desktop/scratch/funny*" )
    ['/Users/dcortes1/Desktop/scratch/funny[file].name']
    >>> glob.glob( "/Users/dcortes1/Desktop/scratch/funny[file].name" )
    []
    >>> glob.glob( "/Users/dcortes1/Desktop/scratch/funny[file].*" )
    []

**Comment by htgoebel** [comment]
@StuntsPT Your glob-pattern is wrong, see https://docs.python.org/2/library/glob.html and https://docs.python.org/2/library/fnmatch.html

@tallforasmurf Thanks for evaluation this.

**Comment by StuntsPT** [comment]
Ok, thanks for looking into this.

**Comment by htgoebel** [comment]
@StuntsPT @tallforasmurf Ups, seems as I was wrong, this is *our* code failing with this pattern.
@StuntsPT How is this pattern created? Please post your .spec-file.

**Comment by StuntsPT** [comment]
Here is my "structure.spec"

```python
# -*- mode: python -*-

block_cipher = None


a = Analysis(['structure.py'],
             pathex=['.', './vars'],
             binaries=None,
             datas=None,
             hiddenimports=['vars.admixprop', 'vars.allelefreq', 'vars.utils', 'vars.marglikehood', 'scipy.special', 'scipy.optimize'],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          name='fastStructure',
          debug=False,
          strip=False,
          upx=True,
          console=True )
```

This fails with the mentioned error on my laptop (ArchLinux, current), but works on my headless server (Ubuntu 12.04), and I get a fully working binary.
I have a lot of files that are related to themes on my laptop, but these are non-existant on the headless server.
I'm trying to build a binary for [this program](https://github.com/rajanil/fastStructure).

**Comment by htgoebel** [comment]
I wonder where this filename or glob-pattern comes from, and why is the theme searched in ...Firefox.. at all. Can you please run pyinstaller under debugger-control (`python -m pdb pyinstaller.py ...`), walk up the stack and try to find which hook is currently executed.

**Comment by StuntsPT** [comment]
@htgoebel 
I have no idea why the theme is being searched at all. This is not a GUI program (albeit it uses scipy, and scipy has some built-in GUI elements AFAIK).
But anyway, here is what I got from using pdb.
**Disclaimer:** This was my first time using pdb like this (I have used it previously inside Eric IDE, but that was it).
The error seems to be in `building/utils.py` (but that we could also get from the backtrace).
Line 454 is where the exception gets raised.
so, from pdb i did:
```python
(Pdb) print(src_root_path_or_glob)
/usr/share/themes/Firefox/Vertex/chrome/images/arrow-dn[dropmarker].png
(Pdb) glob.glob(src_root_path_or_glob)
[]
(Pdb) len(binaries_or_datas)
15914
```

So, from here it indeed looks like and issue with `glob.glob()`, which is returning `[]` when encountering unusual characters, such as "[" or "]", and you were correct all along.
Was this information what you were looking for?
Or was I supposed to do something else? I'm not sure how I can use pdb to "walk up the stack and try to find which hook is currently executed".

**Comment by choidamdam** [comment]
@StuntsPT 
Same issue here: https://github.com/pyinstaller/pyinstaller/issues/1928#issuecomment-218954584 . Maybe my ad-hoc method can help you.

**Comment by StuntsPT** [comment]
Thanks @choidamdam ! I'll try this as soon as I have the chance. It should reduce the size of the generated binaries considerably.

**Comment by htgoebel** [comment]
This was already fixed in 65927314e09e2fbadcf282e5f9881b6f231e2b0e.
